From 5a8af46874795632679a35be1cd3d73d8b338cde Mon Sep 17 00:00:00 2001
From: Stuart Menefy <stuart.menefy@codasip.com>
Date: Tue, 23 Jan 2024 11:03:31 +0000
Subject: [PATCH 02/19] hw/sd/sdcard: Fix response to CMD0 (GO_IDLE_STATE)

This fixes a problem seen when U-Boot is loaded by FSBL:
 - FSBL has just read data from the SD card, leaving the card in
   transfer state.
 - U-Boot issues CMD0 to switch the card into SPI mode.
 - SD card code returns state _before_ the current command, ie transfer
   state.
 - SSI adaptor code translates this to card is _not_ idle.
This causes U-Boot to fail to initialise the card.

Assuming CMD0 works, then the card is now idle, so ignore the
previous state returned by SD card.

Signed-off-by: Stuart Menefy <stuart.menefy@codasip.com>
---
 hw/sd/ssi-sd.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/hw/sd/ssi-sd.c b/hw/sd/ssi-sd.c
index 0b40cf98e9..35b2a4f0d3 100644
--- a/hw/sd/ssi-sd.c
+++ b/hw/sd/ssi-sd.c
@@ -214,6 +214,15 @@ static uint32_t ssi_sd_transfer(SSIPeripheral *dev, uint32_t val)
                 }
 
                 cardstatus = ldl_be_p(longresp);
+
+                /*
+                 * CMD0 resets the card to idle state, so we don't care about
+                 * the state before the command.
+                 */
+                if (s->cmd == 0) {
+                    cardstatus &= ~CURRENT_STATE;
+                }
+
                 status = 0;
                 if (((cardstatus >> 9) & 0xf) < 4)
                     status |= SSI_SDR_IDLE;
-- 
2.43.0

